include/vara/Feature/Constraint.h

```diff
@@ -0,0 +1,379 @@
+#ifndef VARA_FEATURE_CONSTRAINT_H
+#define VARA_FEATURE_CONSTRAINT_H
+
+#include "vara/Feature/Feature.h"
+
+#include "llvm/ADT/StringRef.h"
+
+#include <cassert>
+#include <memory>
+#include <sstream>
+#include <string>
+#include <utility>
+#include <variant>
+
+namespace vara::feature {
+
+//===----------------------------------------------------------------------===//
+//                               Constraint
+//===----------------------------------------------------------------------===//
+
+class ConstraintVisitor;
+
+/// \brief Tree like representation of constraints between features
+class Constraint {
+public:
+  enum class ConstraintKind { CK_BINARY, CK_UNARY, CK_PRIMARY };
+
+  Constraint(ConstraintKind Kind) : Kind(Kind) {}
+  Constraint(const Constraint &) = delete;
+  Constraint &operator=(const Constraint &) = delete;
```

Comment:
```suggestion
  Constraint &operator=(const Constraint &) = delete;
  Constraint(Constraint &&) = delete;
  Constraint &operator=(Constraint &&) = delete;
```
If we don't implement massign and mctor our selfs they will be disabled, so we should at least flag them as deleted here, or implement them.